osu.Game.Rulesets.Tau/UI/TauHardButtonTouchInputMapper.cs

```diff
@@ -0,0 +1,206 @@
+								Height = 1 - normal_area_height_ratio,
+							},
+						]
+					},
+				],
+			},
+		];
+	}
+
+	protected override bool OnKeyDown(KeyDownEvent e)
+	{
+		// Hide whenever the keyboard is used.
+		Hide();
+		return false;
+	}
+
+	protected override bool OnTouchDown(TouchDownEvent e)
+	{
+		return updateAction(e.Touch.Source, getTauActionFromInput(e.ScreenSpaceTouch.Position));
+	}
+
+	protected override void OnTouchMove(TouchMoveEvent e)
+	{
+		updateAction(e.Touch.Source, getTauActionFromInput(e.ScreenSpaceTouch.Position));
+		base.OnTouchMove(e);
+	}
+
+	protected override void OnTouchUp(TouchUpEvent e)
+	{
+		updateAction(e.Touch.Source, null);
+		base.OnTouchUp(e);
+	}
+
+	private bool updateAction(object source, TauAction? newAction)
+	{
+		TauAction? actionBefore = null;
+
+		if (trackedActionSources.TryGetValue(source, out TauAction found))
+			actionBefore = found;
+
+		if (actionBefore != newAction)
+		{
+			if (newAction != null)
+				trackedActionSources[source] = newAction.Value;
+			else
+				trackedActionSources.Remove(source);
+
+			updatePressedActions();
+		}
+
+		return newAction != null;
+	}
+
+	private void updatePressedActions()
+	{
+		Show();
+
+		if (trackedActionSources.ContainsValue(TauAction.HardButton1))
+			keyBindingContainer.TriggerPressed(TauAction.HardButton1);
+		else
+			keyBindingContainer.TriggerReleased(TauAction.HardButton1);
+
+		if (trackedActionSources.ContainsValue(TauAction.HardButton2))
+			keyBindingContainer.TriggerPressed(TauAction.HardButton2);
+		else
+			keyBindingContainer.TriggerReleased(TauAction.HardButton2);
+	}
+
+	private TauAction? getTauActionFromInput(Vector2 screenSpaceInputPosition)
+	{
+		if (hardButton1.Contains(screenSpaceInputPosition))
+			return TauAction.HardButton1;
+		if (hardButton2.Contains(screenSpaceInputPosition))
+			return TauAction.HardButton2;
+		return null;
+	}
+
+	protected override void PopIn() => mainContent.FadeIn(300, Easing.OutQuint);
+
+	protected override void PopOut() => mainContent.FadeOut(300, Easing.OutQuint);
+
+	private partial class InputArea : CompositeDrawable, IKeyBindingHandler<TauAction>
+	{
+		private readonly TauAction handledAction;
+
+		private readonly Box highlightOverlay;
+
+		private readonly IEnumerable<KeyValuePair<object, TauAction>> trackedActions;
+
+		private bool isHighlighted;
+
+		public InputArea(TauAction handledAction, IEnumerable<KeyValuePair<object, TauAction>> trackedActions)
+		{
+			this.handledAction = handledAction;
+			this.trackedActions = trackedActions;
+
+			InternalChildren =
+			[
+				new Container
+				{
+					RelativeSizeAxes = Axes.Both,
+					Masking = true,
+					CornerRadius = 10,
+					Children =
+					[
+						new Box
+						{
+							RelativeSizeAxes = Axes.Both,
+							Alpha = 0.15f,
+						},
+						highlightOverlay = new Box
+						{
+							RelativeSizeAxes = Axes.Both,
+							Alpha = 0,
+							Blending = BlendingParameters.Additive,
+						}
+					]
+				}
+			];
```

Comment:
You always want to use `InternalChild` or `Child` if you're assigning one children element.